modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -216,104 +211,55 @@ public synchronized void stop() throws Exception {
-                    throw e;
-                }
-            }
+        if (schema == null) {
+            throw new IgniteInternalException(format("Schema not found [schemaName={}]", schemaName));
         }
 
-        SqlNodeList qryList = Commons.parse(sql, FRAMEWORK_CONFIG.getParserConfig());
-        List<SqlCursor<List<?>>> cursors = new ArrayList<>(qryList.size());
-
-        List<RootQuery<Object[]>> qrys = new ArrayList<>(qryList.size());
-
-        for (final SqlNode sqlNode : qryList) {
-            RootQuery<Object[]> qry = new RootQuery<>(
-                    sqlNode.toString(),
-                    schemaManager.schema(schemaName), // Update schema for each query in multiple statements.
-                    params,
-                    exchangeService,
-                    (q) -> queryRegistry.unregister(q.id()),
-                    LOG
-            );
-
-            qrys.add(qry);
-
-            queryRegistry.register(qry);
-
-            try {
-                if (qryList.size() == 1) {
-                    plan = planCache.queryPlan(
-                            new CacheKey(schemaName, qry.sql()),
-                            () -> prepareSvc.prepareSingle(sqlNode, qry.planningContext()));
-                } else {
-                    plan = prepareSvc.prepareSingle(sqlNode, qry.planningContext());
-                }
-
-                final QueryPlan finalPlan = plan;
-
-                context.maybeUnwrap(QueryValidator.class)
-                        .ifPresent(queryValidator -> queryValidator.validatePlan(finalPlan));
-
-                cursors.add(executionSrvc.executePlan(qry, plan));
-            } catch (Exception e) {
-                boolean isCanceled = qry.isCancelled();
-
-                qrys.forEach(RootQuery::cancel);
-
-                queryRegistry.unregister(qry.id());
-
-                if (isCanceled) {
-                    throw new IgniteInternalException("The query was cancelled while planning", e);
-                } else {
-                    throw e;
-                }
-            }
+        SqlNodeList nodes = parserCache.computeIfAbsent(sql, key -> Commons.parse(key, FRAMEWORK_CONFIG.getParserConfig()));
+
+        var res = new ArrayList<CompletableFuture<AsyncSqlCursor<List<Object>>>>(nodes.size());
+
+        CompletableFuture<Void> start = new CompletableFuture<>();
+
+        for (SqlNode sqlNode : nodes) {
+            BaseQueryContext ctx = BaseQueryContext.builder()
+                    .cancel(new QueryCancel())
+                    .frameworkConfig(
+                            Frameworks.newConfigBuilder(FRAMEWORK_CONFIG)
+                                    .defaultSchema(schema)
+                                    .build()
+                    )
+                    .logger(LOG)
+                    .parameters(params)
+                    .build();
+
+            CompletableFuture<AsyncSqlCursor<List<Object>>> stage = start.thenCompose(voidArg -> prepareSvc.prepare(sqlNode, ctx))
```

Comment:
What is the idea behind `start` future, can we just use the result of `prepareSvc.prepare` call?

Reply:
The idea is to avoid doing the query-related tasks on the calling thread. In the case when preparation will be completed before the rest of chain will be build, the tail will be executed on the calling thread. 

You could learn more from [this](https://youtu.be/W7iK74YA5NM?t=2208) video.